### ISSUE TITLE
doc localhost vs 127.0.0.1 for test services

### ISSUE THREAD

**Issue by drewcoo** [circleci/circleci-docs#2487]
# Missing Information
I don't know whether this is a bug or missing documentation so I'm going with documentation.
Apparently I should run my test service on 127.0.0.1:test_port instead of localhost:test_port. Because that made my tests pass on Circle.
Here's an example of them failing: https://circleci.com/gh/drewcoo/doppelserver/38
They're set to 127.0.0.1 and pass on the next build.

# Already Looked
Searched through these docs and your main issue tracker.
I found this, which has examples of localhost which misled me.
https://circleci.com/docs/2.0/databases/#using-dockerize-to-wait-for-dependencies

# Potential Locations
It doesn't matter to me where it's doc's but it would be nice if there were some searchable place that hit on some subset of following key terms:
- localhost
- 127.0.0.1
- port
- TCP connection
- requested address
- assign

Or maybe this is a bug. Let me know and I'll fix that.

**Comment by levlaz** [comment]
I think the problem might be that you were specifying `ports` in one build and not the other. If you switch back to `localhost` in the passing build without specifying `ports` do you get the same error?

**Comment by drewcoo** [comment]
The problem has disappeared. But it did exist.

The code ran successfully on my machine and on TravisCI, Switching from localhost to 127.0.0.1 fixed the problem at the time.

But I cannot repro now. I even have a branch just for this purpose and I succeed when set to localhost:
https://ci.appveyor.com/project/drewcoo/doppelserver/build/1.0.32-appveyor_repro

I guess we can close this but not understanding it bothers me.

**Comment by drewcoo** [comment]
Here you go:
https://circleci.com/gh/drewcoo/doppelserver/tree/circleci_repro

The first build on that branch (#43) is set to 127.0.0.1 and succeeds.
The second build on that branch (#44) is set to localhost. That's the only change as you can see in the commit log:
https://github.com/drewcoo/doppelserver/compare/6fd5eb6a933b...4ed10b54eb94
And it fails.

Let me know if there's anything else I can do to help diagnose.

**Comment by smart-alek** [comment]
@drewcoo sorry for the wait on this. Based on the [configuration reference docs](https://circleci.com/docs/2.0/configuration-reference/#docker), I wonder if you have to define a service container under the `docker` key before anything gets exposed on `localhost`.

In both of your config files, you've got a postgres image defined but commented out. Of course, this doesn't explain why pointing to 127.0.0.1 succeeds for you, but it made me pause, at least. 🤔

**Comment by drewcoo** [comment]
No rush. I set aside a branch with a locahost repro for you in case it's a real problem and I'm using 127.0.0.1 for my other code. I'm not blocked.

The only thing I changed before repushing the code was the server difference. You can see it in the source diff:
https://github.com/drewcoo/doppelserver/commit/4ed10b54eb947bc640baeefebff6deb19653be85
I didn't change anything outside of source control when I put the repro together.

I'll spend some time tonight or tomorrow seeing if I can find any gotchas with Docker.

**Comment by smart-alek** [comment]
Hey @drewcoo, just confirmed that this is indeed a real issue. Can you try nuking the second line of that box's `/etc/hosts`? Trying it over SSH is probably simplest.

**Comment by drewcoo** [comment]
So I can ssh in but I can't write to /etc/hosts. I had an altered version of hosts and tried 'sudo mv' to overwrite I was told "mv: cannot move 'my_hosts' to '/etc/hosts': Device or resource busy" which makes sense.

Frankly, I'm wondering if it's some Ruby library code at fault. Maybe I'm using something that only understands ipv4 and it's talking to something that's giving it an ipv6 address, so when it asks for localhost expecting ipv4 it gets ::1 and freaks out.

I'll apologize right now just in case that's the problem. It didn't occur to me. Let me dig a little more to see if that's it.

**Comment by smart-alek** [comment]
Yeah, your suspicion makes sense to me, but no apologies necessary! You aren't the first to stumble into this, so we're taking it seriously. Let me know how your Ruby library spelunking goes, and we can close this issue down if you find something there. :)

**Comment by drewcoo** [comment]
Hang on. Other people have hit this? What information do we have about them?

Also, what do I know? I know what Docker container and Ruby version and versions of gems I'm using. And I know that whatever is wrong it's not that way locally or on Travis. I'll check that versions of everything match and then look at environment differences.

I almost let this slip again. Please do me a favor and ping me at the beginning of next week if I don't get back to this thread.

**Comment by smart-alek** [comment]
Other people have hit this in the past, but you're actually the first in 2018. :) I don't have deets on them because of how far back it was — our support team is just wrapping their heads around it.

I'll ping you next week to see what you've found and provide an update on our end! 👍

**Comment by smart-alek** [comment]
And a bit of info for you from the 2017 cases we saw: we were able to fix this by appending `127.0.0.1 localhost` to `/etc/hosts`. The unfortunate business occurs when ipv6 appears last instead of ipv4.

But! This gets tricky because you've said you can't modify your `etc/hosts`. Stand by for more info on our end. :)

**Comment by smart-alek** [comment]
@drewcoo absolutely no worries! It's not a waste of time — it's my job! 🙃 

Thanks for **your** patience and seeing this through — I'll relay your findings back to the rest of the team. Bon voyage on your coding journeys! 🚢
